### PR TITLE
Model Deployment 

### PR DESCRIPTION
**Pull request by nfranczak**
This PR:

Following along with [Marco's PR](https://github.com/viamrobotics/rdk/pull/1052), this PR aims to allow a user to deploy a model stored in Viam's cloud onto their local robot.

To test this functionality, we have added `TestModelDeploy()` within `builtin_test.go`. 

NOTE: This PR is a continue is a continuation of this [former PR](https://github.com/viamrobotics/rdk/pull/1200).

### CODE REVIEW

File: services/datamanager/builtin/builtin.go [hunk 1]
```diff
@@ -430,8 +435,29 @@ func (svc *builtIn) Update(ctx context.Context, cfg *config.Config) error {
...
+			svc.modelManager = modelManager
+		}
+
+		// Download models from models_on_robot.
+		modelsToDeploy := svcConfig.ModelsToDeploy
+		errorChannel := make(chan error, len(modelsToDeploy))
+		go svc.modelManager.DownloadModels(cfg, modelsToDeploy, errorChannel)
+		if len(errorChannel) != 0 {
+			var s string
+			for i := 0; i < len(errorChannel); i++ {
```
Comment (nfranczak) [inline services/datamanager/builtin/builtin.go:455]: is now just:
```
if len(errorChannel) != 0 {
    var s string
    for i := range errorChannel {
        s = s + "\n" + i.Error()
    }
    return errors.New(s)
}
```

File: services/datamanager/model/model.go [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+	m.cancelFunc()
+	if m.conn != nil {
+		if err := m.conn.Close(); err != nil {
+			m.logger.Errorw("error closing model deploy server connection", "error", err)
+		}
+	}
+}
+
+// DownloadModels handles deploying models into their specified destination.
+func (m *modelManager) DownloadModels(cfg *config.Config, modelsToDeploy []*Model, errorChannel chan error) chan error {
```
Comment (agreenb) [inline services/datamanager/model/model.go:119]: I may have missed previous context, but I think this should not return anything. The error channel passed in is to ensure that when this is concurrent, the caller actually receives the error and can act on it.

File: services/datamanager/model/model.go [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+	for _, model := range modelsToDownload {
+		deployRequest := &v1.DeployRequest{
+			Metadata: &v1.DeployMetadata{
+				ModelName: model.Name,
+			},
+		}
+		deployResp, err := m.deploy(cancelCtx, deployRequest)
+		if err != nil {
+			m.logger.Error(err)
+			errorChannel <- err
```
Comment (agreenb) [inline services/datamanager/model/model.go:145]: If there is an error here, `deployResponse` will likely be nil and so would want to return here.

File: services/datamanager/model/model.go [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+		}
+		url := deployResp.Message
+		filePath := filepath.Join(model.Destination, model.Name)
+		err = downloadFile(cancelCtx, m.httpClient, filePath, url, m.logger)
+		if err != nil {
+			m.logger.Error(err)
+			errorChannel <- err
```
Comment (agreenb) [inline services/datamanager/model/model.go:152]: If there is an error here, the file has not been downloaded so there won't be anything to unzip at filepath. Can return after throwing the error into the channel.

File: services/datamanager/model/model.go [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+		}
+		// A download from a GCS signed URL only returns one file.
+		modelFileToUnzipPath := filepath.Join(model.Destination, model.Name+zipExtension)
+		if err = unzipSource(modelFileToUnzipPath, model.Destination); err != nil {
+			m.logger.Error(err)
+			errorChannel <- err
+		}
+	}
+	close(errorChannel)
```
Comment (agreenb) [inline services/datamanager/model/model.go:161]: This close can happen by the caller, and instead don't return anything here.

File: services/datamanager/model/model.go [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+			// Set the model destination to default if it's not specified in the config.
+			model.Destination = filepath.Join(ViamModelDotDir, model.Name)
+		}
+		_, err := os.Stat(model.Destination)
+		switch {
+		case errors.Is(err, os.ErrNotExist):
+			// we know we have never attempted to deploy the model
+			modelsToDownload = append(modelsToDownload, model)
+			// create model.Destination directory
+			err := os.MkdirAll(model.Destination, os.ModePerm)
```
Comment (agreenb) [inline services/datamanager/model/model.go:182]: can shorten this to:
```
if err := os.MkdirAll(model.Destination, os.ModePerm); err != nil {
  return nil, err
}
```

File: services/datamanager/model/model.go [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+	for _, model := range models {
+		if model.Destination == "" {
+			// Set the model destination to default if it's not specified in the config.
+			model.Destination = filepath.Join(ViamModelDotDir, model.Name)
+		}
+		_, err := os.Stat(model.Destination)
+		switch {
+		case errors.Is(err, os.ErrNotExist):
+			// we know we have never attempted to deploy the model
+			modelsToDownload = append(modelsToDownload, model)
```
Comment (agreenb) [inline services/datamanager/model/model.go:180]: supernit: can add `modelsToDownload` right after trying to make the model.Destination directory instead of right before, since no need to append if erroring out

File: services/datamanager/model/model.go [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+			}
+		case err != nil:
+			return nil, err
+		default:
+			files, err := ioutil.ReadDir(model.Destination)
+			if err != nil {
+				return nil, err
+			}
+			if len(files) == 0 {
+				// know there was a partial download
```
Comment (agreenb) [inline services/datamanager/model/model.go:194]: Could you expand slightly here? The model destination directory was created but there weren't any files, so there was a partial download?

**Review by agreenb — CHANGES_REQUESTED** [review]
Nice work iterating on this! Great changes to have an end-to-end integration test, refactor with builtin, and remove concurrency for now. Relatively minor changes then good to go for this PR!

File: services/datamanager/builtin/builtin.go [hunk 9]
```diff
@@ -430,8 +435,28 @@ func (svc *builtIn) Update(ctx context.Context, cfg *config.Config) error {
...
+		}
+
+		// Download models from models_on_robot.
+		modelsToDeploy := svcConfig.ModelsToDeploy
+		errorChannel := make(chan error, len(modelsToDeploy))
+		go svc.modelManager.DownloadModels(cfg, modelsToDeploy, errorChannel)
+		if len(errorChannel) != 0 {
+			var s string
+			for i := range errorChannel {
+				s = s + "\n" + i.Error()
```
[the remark this answers is not in the document]
Comment (agreenb) [reply services/datamanager/builtin/builtin.go:456]: +1

[the remark this answers is not in the document]
Comment (nfranczak) [reply services/datamanager/builtin/builtin.go:456]: changed.

File: services/datamanager/builtin/builtin_test.go [hunk 10]
```diff
@@ -332,6 +339,105 @@ func TestRecoversAfterKilled(t *testing.T) {
 	test.That(t, len(mockService.getUploadedFiles()), test.ShouldEqual, 1+numArbitraryFilesToSync)
 }
 
+// TODO(DATA-341): Handle partial downloads in order to resume deployment.
+// TODO(DATA-344): Compare checksum of downloaded model to blob to determine whether to redeploy.
+// Want to make sure user can deploy to a custom location
```
[the remark this answers is not in the document]
Comment (nfranczak) [reply services/datamanager/builtin/builtin_test.go:344]: removed.

File: services/datamanager/builtin/builtin_test.go [hunk 11]
```diff
@@ -332,6 +339,105 @@ func TestRecoversAfterKilled(t *testing.T) {
...
+// Validates that models can be deployed onto a robot.
+func TestModelDeploy(t *testing.T) {
+	deployedZipFileName := "model.zip"
```
[the remark this answers is not in the document]
Comment (nfranczak) [reply services/datamanager/builtin/builtin_test.go:347]: thanks! :)

File: services/datamanager/builtin/builtin_test.go [hunk 12]
```diff
@@ -332,6 +339,105 @@ func TestRecoversAfterKilled(t *testing.T) {
...
+	// Add content to zip file.
+	b0 := []byte("text representing model.txt internals.")
+	r0 := bytes.NewReader(b0)
+	b1 := []byte("text representing README.md internals.")
+	r1 := bytes.NewReader(b1)
+	zipWriter := zip.NewWriter(zipper)
+
+	zipFile1, err := zipWriter.Create(originalFileName)
+	test.That(t, err, test.ShouldBeNil)
+	_, err = io.Copy(zipFile1, r0)
```
[the remark this answers is not in the document]
Comment (nfranczak) [reply services/datamanager/builtin/builtin_test.go:363]: yes, using `zipFile1.Write(b0)` works! Changed now.

File: services/datamanager/builtin/builtin_test.go [hunk 13]
```diff
@@ -332,6 +339,105 @@ func TestRecoversAfterKilled(t *testing.T) {
...
+	test.That(t, err, test.ShouldBeNil)
+
+	zipFile2, err := zipWriter.Create(otherOriginalFileName)
+	test.That(t, err, test.ShouldBeNil)
+	_, err = io.Copy(zipFile2, r1)
+	test.That(t, err, test.ShouldBeNil)
+
+	// Now that we are done editing the zip file we close.
+	zipWriter.Close()
```
[the remark this answers is not in the document]
Comment (nfranczak) [reply services/datamanager/builtin/builtin_test.go:372]: ok. Now we have
```
	zipper, err := os.Create(deployedZipFileName)
	test.That(t, err, test.ShouldBeNil)
	defer zipper.Close()
```
However I cannot do:
```
	zipWriter := zip.NewWriter(zipper)
	defer zipWriter.Close()
```
When I `defer zipWriter.Close()`, I error at `zip.OpenReader(fileNamePath)` inside `model/unzipSource()`.

After digging around I realized that if we have `defer zipWriter.Close()` then the reason why we error  inside `model/unzipSource()` is because the zipfiles themselves do not contain any data - they are 0kB. This means that there is something wrong with `func (m *mockClient) Do`. However the code in that function is correct..? inside that function we have `buf, err := os.ReadFile(m.zipFileName)` but we cannot read a file if it has not been closed.

File: services/datamanager/builtin/builtin_test.go [hunk 14]
```diff
@@ -332,6 +339,105 @@ func TestRecoversAfterKilled(t *testing.T) {
...
+	dmCfg, err := getDataManagerConfig(testCfg)
+	test.That(t, err, test.ShouldBeNil)
+
+	// Set SyncIntervalMins equal to zero so we do not enable syncing.
+	dmCfg.SyncIntervalMins = 0
+	dmCfg.ModelsToDeploy = allModels
+
+	// Initialize the data manager and update it with our config.
+	dmsvc := newTestDataManager(t, "arm1", "")
+	dmsvc.SetModelManagerConstructor(getTestModelManagerConstructor(t, modelServer, deployedZipFileName))
```
[the remark this answers is not in the document]
Comment (nfranczak) [reply services/datamanager/builtin/builtin_test.go:408]: done. The ticket is now

File: services/datamanager/builtin/builtin_test.go [hunk 15]
```diff
@@ -332,6 +339,105 @@ func TestRecoversAfterKilled(t *testing.T) {
...
+	// Validate that the models were deployed.
+	files, err := ioutil.ReadDir(allModels[0].Destination)
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, len(files), test.ShouldEqual, 2)
+	files, err = ioutil.ReadDir(allModels[1].Destination)
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, len(files), test.ShouldEqual, 2)
+
+	// Validate that the deployed model file equals the dummy file that was zipped.
+	deployedUnzippedFile, err := ioutil.ReadFile(filepath.Join(allModels[0].Destination, originalFileName))
```
[the remark this answers is not in the document]
Comment (nfranczak) [reply services/datamanager/builtin/builtin_test.go:427]: done. we now have `builtin_test.go/fileCompareTestHelper()`

File: services/datamanager/model/model.go [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+func (m *modelManager) Close() {
+	m.cancelFunc()
+	if m.conn != nil {
+		if err := m.conn.Close(); err != nil {
+			m.logger.Errorw("error closing model deploy server connection", "error", err)
+		}
+	}
+}
+
+// DownloadModels handles deploying models into their specified destination.
```
[the remark this answers is not in the document]
Comment (nfranczak) [reply services/datamanager/model/model.go:118]: changed.

File: services/datamanager/model/model.go (hunk 2 above)
Comment (nfranczak) [reply services/datamanager/model/model.go:119]: gotcha, we now return nothing

File: services/datamanager/model/model.go (hunk 3 above)
Comment (nfranczak) [reply services/datamanager/model/model.go:145]: gotcha, added in.

File: services/datamanager/model/model.go (hunk 4 above)
Comment (nfranczak) [reply services/datamanager/model/model.go:152]: gotcha, added in.

File: services/datamanager/model/model.go (hunk 5 above)
Comment (nfranczak) [reply services/datamanager/model/model.go:161]: removed `close(errorChannel)` . now it does not appear anywhere. Is this what you meant?

File: services/datamanager/model/model.go (hunk 7 above)
Comment (nfranczak) [reply services/datamanager/model/model.go:180]: agreed.

File: services/datamanager/model/model.go (hunk 6 above)
Comment (nfranczak) [reply services/datamanager/model/model.go:182]: shortened.

File: services/datamanager/model/model.go (hunk 8 above)
Comment (nfranczak) [reply services/datamanager/model/model.go:194]: Was thinking ahead wrt partial downloads. Not needed in this version of the code I have removed it.

File: services/datamanager/builtin/builtin.go [hunk 17]
```diff
@@ -450,10 +450,11 @@ func (svc *builtIn) Update(ctx context.Context, cfg *config.Config) error {
 		modelsToDeploy := svcConfig.ModelsToDeploy
 		errorChannel := make(chan error, len(modelsToDeploy))
 		go svc.modelManager.DownloadModels(cfg, modelsToDeploy, errorChannel)
+		// close(errorChannel)
```
Comment (agreenb) [inline services/datamanager/builtin/builtin.go:453]: Can remove here.

From a SO answer / Go book:

> You needn't close every channel when you've finished with it. It's only necessary to close a channel when it is important to tell the receiving goroutines that all data have been sent. A channel that the garbage collector determines to be unreachable will have its resources reclaimed whether or not it is closed. (Don't confuse this with the close operation for open files. It is important to call the Close method on every file when you've finished with it.)

File: services/datamanager/builtin/builtin_test.go [hunk 18]
```diff
@@ -350,27 +349,26 @@ func TestModelDeploy(t *testing.T) {
...
 	zipper, err := os.Create(deployedZipFileName)
 	test.That(t, err, test.ShouldBeNil)
+	defer zipper.Close()
 
 	// Add content to zip file.
 	b0 := []byte("text representing model.txt internals.")
-	r0 := bytes.NewReader(b0)
 	b1 := []byte("text representing README.md internals.")
-	r1 := bytes.NewReader(b1)
 	zipWriter := zip.NewWriter(zipper)
```
Comment (agreenb) [inline services/datamanager/builtin/builtin_test.go:357]: Supernit: could you move b0/b1 instantiation after the file names, then create the zipper, then create the zipwriter + write? Follows a bit of a more natural ordering flow.

File: services/datamanager/builtin/builtin_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (agreenb) [reply services/datamanager/builtin/builtin_test.go:371]: I believe the fix may be:
```
	// Add content to zip file.
	b0 := []byte("text representing model.txt internals.")
	r0 := bytes.NewReader(b0)
	b1 := []byte("text representing README.md internals.")
	r1 := bytes.NewReader(b1)

        // Create zip file.
	deployedZipFile, err := os.Create(deployedZipFileName)
	test.That(t, err, test.ShouldBeNil) 
	zipWriter := zip.NewWriter(zipper)

	defer deployedZipFile.Close()
	defer zipWriter.Close()
```
Defers happen in a LIFO (last in first out) order. 

_but_ after reading around, https://www.joeshaw.org/dont-defer-close-on-writable-files/ indicates that you should explicitly close files rather than deferring, and just make sure to close on any errors.

File: services/datamanager/builtin/builtin_test.go [hunk 19]
```diff
@@ -803,6 +813,8 @@ type mockDataSyncServiceServer struct {
 	v1.UnimplementedDataSyncServiceServer
 }
 
+// TODO(DATA-487): Support zipping multiple files in
```
Comment (agreenb) [inline services/datamanager/builtin/builtin_test.go:816]: Thank you for adding

File: services/datamanager/builtin/builtin.go [hunk 20]
```diff
@@ -450,10 +450,11 @@ func (svc *builtIn) Update(ctx context.Context, cfg *config.Config) error {
...
 		if len(errorChannel) != 0 {
 			var s string
-			for i := range errorChannel {
-				s = s + "\n" + i.Error()
+			for err := range errorChannel {
+				s = s + "," + err.Error()
```
Comment (agreenb) [inline services/datamanager/builtin/builtin.go:457]: This would cause a comma to prepend the initial error. Adding it after would add one-too-many commas at the end.
So instead you could do:

```
var errMsgs []string
for err := range errorChannel {
   errMsgs = append(errMsgs, err.Error()
}
return Errors.New(strings.Join(errMsgs[:], ", "))
```
If it's empty, it returns an empty string. And then joins with ", " for len(errMsgs) > 1

**Review by agreenb — CHANGES_REQUESTED** [review]
Just a few last comments, nice work.

File: services/datamanager/builtin/builtin_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (nfranczak) [reply services/datamanager/builtin/builtin_test.go:371]: hmm does not work :/.. still throws the same error. I will stick with the explicit close.

File: services/datamanager/builtin/builtin.go (hunk 17 above)
Comment (nfranczak) [reply services/datamanager/builtin/builtin.go:454]: removed.

File: services/datamanager/builtin/builtin.go (hunk 20 above)
Comment (nfranczak) [reply services/datamanager/builtin/builtin.go:458]: changed.

File: services/datamanager/builtin/builtin_test.go (hunk 18 above)
Comment (nfranczak) [reply services/datamanager/builtin/builtin_test.go:357]: gotcha.

File: services/datamanager/builtin/builtin_test.go [hunk 21]
```diff
@@ -332,6 +339,105 @@ func TestRecoversAfterKilled(t *testing.T) {
...
+	zipFile2, err := zipWriter.Create(otherOriginalFileName)
+	test.That(t, err, test.ShouldBeNil)
+	_, err = io.Copy(zipFile2, r1)
+	test.That(t, err, test.ShouldBeNil)
+
+	// Now that we are done editing the zip file we close.
+	zipWriter.Close()
+	zipper.Close()
+
+	defer os.Remove(deployedZipFileName)
```
[the remark this answers is not in the document]
Comment (nfranczak) [reply services/datamanager/builtin/builtin_test.go:373]: done.

File: services/datamanager/builtin/builtin_test.go (hunk 19 above)
Comment (nfranczak) [reply services/datamanager/builtin/builtin_test.go:816]: np!

File: services/datamanager/builtin/builtin_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (agreenb) [reply services/datamanager/builtin/builtin_test.go:373]: The reason that deferring `zipWriter.Close()` caused the issue with 0kb data is that Close finishes writing the zip file by writing the central directory, and does not close the actual underlying writer (which is why you still call `deployedZipFile.Close()`). Based on that, the current explicit close after writing is correct.

File: services/datamanager/builtin/builtin_test.go [hunk 22]
```diff
@@ -422,7 +420,7 @@ func TestModelDeploy(t *testing.T) {
 	test.That(t, err, test.ShouldBeNil)
 	test.That(t, len(files), test.ShouldEqual, 2)
 
-	// Validate that the deployed model file equals the dummy file that was zipped.
+	// Validate that the deployed model files equal the dummy files that was zipped.
```
Comment (agreenb) [inline services/datamanager/builtin/builtin_test.go:423]: supernit: 'that were zipped'

File: services/datamanager/builtin/builtin.go [hunk 23]
```diff
@@ -431,8 +437,28 @@ func (svc *builtIn) Update(ctx context.Context, cfg *config.Config) error {
 		svc.closeCollectors()
 		return err
 	}
-	if cfg.Cloud != nil {
-		svc.partID = cfg.Cloud.ID
+
+	// Check that we have models to download and appropriate credentials.
+	if svcConfig.ModelsToDeploy != nil && cfg.Cloud != nil {
```
Comment (agreenb) [inline services/datamanager/builtin/builtin.go:442]: Want to instead check if len(svcConfig.ModelsToDeploy) > 0 rather than nil, since if this gets initialized in Update and `models_to_deploy` is set to [], this will not be nil but there still won't be any models to deploy.

Comment (agreenb) [inline services/datamanager/builtin/builtin.go:442]: The previous comment made me also think that we should add a test that directly reads from a new `data_manager_deploy_model.json` to be sure that we're correctly parsing `models_to_deploy` from a config. We inject ModelsToDeploy into the Config format in the current test, but likely need to explicitly add a json parser for Models within `models_to_deploy`

Could you add a TODO in the test for the above? Thank you!

**Review by agreenb — CHANGES_REQUESTED** [review]
Thanks for iterating on this - just a few last small comments then looks good to merge!

File: services/datamanager/builtin/builtin_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (nfranczak) [reply services/datamanager/builtin/builtin_test.go:373]: kk

File: services/datamanager/builtin/builtin.go (hunk 23 above)
Comment (nfranczak) [reply services/datamanager/builtin/builtin.go:442]: done. we now have DATA-493 added above.

File: services/datamanager/builtin/builtin_test.go (hunk 22 above)
Comment (nfranczak) [reply services/datamanager/builtin/builtin_test.go:423]: changed.

**Review by agreenb — APPROVED** [review]
LGTM, thank you for creating and linking the concurrency ticket and skipping this test in the meantime. Appreciate updating the model deployment notes with the current state of the project!

This was a huge PR on a very complex technical problem - nice work iterating through it. 🚀 🚀
